File: aoc15/src/test/java/aoc15/day03/Day03Test.java
package aoc15.day03;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class Day03Test {

  @ParameterizedTest
  @CsvSource({
      ">, 2",
      "^>v<, 4",
      "^v^v^v^v^v^v, 2"
  })
  void part1Test(String input, int expectedResult) {
    var day = new Day03();

    assertThat(day.part1(input))
        .isEqualTo(expectedResult);
  }

  @ParameterizedTest
  @CsvSource({
      "^v, 3",
      "^>v<, 3",
      "^v^v^v^v^v, 11"
  })
  void part2Test(String input, int expectedResult) {
    var day = new Day03();

    assertThat(day.part2(input))
        .isEqualTo(expectedResult);
  }
}

File: aoc25/src/test/java/day03/Day03Test.kt
package day03

import aoc.loadExample
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.Arguments.arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.stream.Stream

class Day03Test {

    val filename = "day03.txt"

    @Test
    fun part1Example() {
        val input = loadExample(filename)
        val subject = Day03()

        val result = subject.part1(input)

        assertThat(result).isEqualTo(357L)
    }

    @Test
    fun part2Example() {
        val input = loadExample(filename)
        val subject = Day03()

        val result = subject.part2(input)

        assertThat(result).isEqualTo(3121910778619)
    }

    @ParameterizedTest
    @MethodSource("getMaxJoltage1Arguments")
    fun getMaxJoltage1WorksCorrectly(bankString: String, expected: Long) {
        val bank = convertBankString(bankString)

        assertThat(getMaxJoltage1(bank)).isEqualTo(expected)
    }

    @ParameterizedTest
    @MethodSource("getMaxJoltage2Arguments")
    fun getMaxJoltage2WorksCorrectly(bankString: String, expected: Long) {
        val bank = convertBankString(bankString)

        assertThat(getMaxJoltage2(bank)).isEqualTo(expected)
    }

    companion object {
        @JvmStatic
        fun getMaxJoltage1Arguments(): Stream<Arguments> {
            return Stream.of(
                arguments("987654321111111", 98),
                arguments("811111111111119", 89),
                arguments("234234234234278", 78),
                arguments("818181911112111", 92),
            )
        }

        @JvmStatic
        fun getMaxJoltage2Arguments(): Stream<Arguments> {
            return Stream.of(
                arguments("987654321111111", 987654321111),
                arguments("811111111111119", 811111111119),
                arguments("234234234234278", 434234234278),
                arguments("818181911112111", 888911112111),
            )
        }
    }

    fun convertBankString(bankString: String): List<Int> =
        bankString.asSequence().map { it.digitToInt() }.toList()
}